avm-transpiler/src/procedures/parser.rs

```diff
@@ -0,0 +1,185 @@
+use crate::instructions::AvmTypeTag;
+use once_cell::sync::Lazy;
+use regex::Regex;
+
+#[allow(clippy::upper_case_acronyms)]
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum Alias {
+    // Compute
+    ADD,
+    SUB,
+    MUL,
+    DIV,
+    FDIV,
+    EQ,
+    LT,
+    LTE,
+    AND,
+    OR,
+    XOR,
+    NOT,
+    SHL,
+    SHR,
+    CAST,
+    // Control flow
+    JUMP,
+    JUMPI,
+    // Memory
+    SET,
+    MOV,
+    // Misc
+    INTERNALRETURN,
+    ECADD,
+    TORADIXBE,
+}
+
+impl Alias {
+    fn lookup(as_string: &str) -> Alias {
+        match as_string {
+            "ADD" => Alias::ADD,
+            "SUB" => Alias::SUB,
+            "MUL" => Alias::MUL,
+            "DIV" => Alias::DIV,
+            "FDIV" => Alias::FDIV,
+            "EQ" => Alias::EQ,
+            "LT" => Alias::LT,
+            "LTE" => Alias::LTE,
+            "AND" => Alias::AND,
+            "OR" => Alias::OR,
+            "XOR" => Alias::XOR,
+            "NOT" => Alias::NOT,
+            "SHL" => Alias::SHL,
+            "SHR" => Alias::SHR,
+            "CAST" => Alias::CAST,
+            "JUMP" => Alias::JUMP,
+            "JUMPI" => Alias::JUMPI,
+            "INTERNALRETURN" => Alias::INTERNALRETURN,
+            "SET" => Alias::SET,
+            "MOV" => Alias::MOV,
+            "ECADD" => Alias::ECADD,
+            "TORADIXBE" => Alias::TORADIXBE,
+            _ => unreachable!("Invalid alias {}", as_string),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct Label {
+    inner: String,
+}
+
+impl Label {
+    fn new(inner: String) -> Self {
+        Label { inner }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Symbol {
+    Direct(usize),
+    Indirect(usize),
+    Reserved(usize),
+    Label(Label),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum Operand {
+    Symbol(Symbol),
+    Immediate(u128),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ParsedOpcode {
+    pub(crate) alias: Alias,
+    pub(crate) operands: Vec<Operand>,
+    pub(crate) tag: Option<AvmTypeTag>,
+    pub(crate) label: Option<Label>,
+}
+
+pub(crate) type Assembly = Vec<ParsedOpcode>;
+
+// Simple regex parser
+pub(crate) fn parse(assembly: &str) -> Result<Assembly, String> {
```

Comment:
maybe call the input something else? since you are calling `Assembly` a vec of opcodes.